File: src/Foreign/Object/ST.lua
return {
  new = (function()
    return {}
  end),
  peekImpl = (function(just, nothing, k, m)
    return function()
      if m[k] == nil then
        return nothing
      else
        return just(m[k])
    end
  end),
  pokeImpl = (function(k, v, m)
    return function()
      m[k] = v
      return m
    end
  end),
  deleteImpl = (function(k, m)
    return function()
      m[k] = nil
      return m
    end
  end),
}

File: src/Foreign/Object/ST.js
const newImpl = function () {
  return {};
};
export { newImpl as new };

export function peekImpl(just, nothing, k, m) {
  return function () {
    return {}.hasOwnProperty.call(m, k) ? just(m[k]) : nothing;
  };
}

export function pokeImpl(k, v, m) {
  return function () {
    m[k] = v;
    return m;
  };
}

export const deleteImpl = function (k, m) {
  return function () {
    delete m[k];
    return m;
  };
};
